mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClustering.scala

```diff
@@ -0,0 +1,631 @@
+    this(center, rows, variances, breezeNorm(variances.toBreeze, 2.0),
+      0.0, None, Array.empty[ClusterTree])
+
+  /**
+   * Inserts a sub node as its child
+   *
+   * @param child inserted sub node
+   */
+  def insert(child: ClusterTree) {
+    insert(Array(child))
+  }
+
+  /**
+   * Inserts sub nodes as its children
+   *
+   * @param children inserted sub nodes
+   */
+  def insert(children: Array[ClusterTree]) {
+    this.children = this.children ++ children
+    children.foreach(child => child.parent = Some(this))
+  }
+
+  /**
+   * Converts the tree into Array class
+   * the sub nodes are recursively expanded
+   *
+   * @return an Array class which the cluster tree is expanded
+   */
+  def toArray(): Array[ClusterTree] = {
+    val array = this.children.size match {
+      case 0 => Array(this)
+      case _ => Array(this) ++ this.children.flatMap(child => child.toArray().toIterator)
+    }
+    array.sortWith { case (a, b) =>
+      a.getDepth < b.getDepth && a.variances.toArray.sum < b.variances.toArray.sum
+    }
+  }
+
+  /**
+   * Gets the depth of the cluster in the tree
+   *
+   * @return the depth from the root
+   */
+  def getDepth: Int = {
+    this.parent match {
+      case None => 0
+      case _ => 1 + this.parent.get.getDepth
+    }
+  }
+
+  /**
+   * Gets the leaves nodes in the cluster tree
+   */
+  def getLeavesNodes: Array[ClusterTree] = {
+    this.toArray().filter(_.isLeaf).sortBy(_.center.toArray.sum)
+  }
+
+  def isLeaf: Boolean = (this.children.size == 0)
+
+  def getParent: Option[ClusterTree] = this.parent
+
+  def getChildren: Seq[ClusterTree] = this.children
+
+  /**
+   * Gets the dendrogram height of the cluster at the cluster tree.
+   * A dendrogram height is different from a local height.
+   * A dendrogram height means a total height of a node in a tree.
+   * A local height means a maximum distance between a node and its children.
+   *
+   * @return the dendrogram height
+   */
+  def getHeight: Double = {
+    this.children.size match {
+      case 0 => 0.0
+      case _ => this.localHeight + this.children.map(_.getHeight).max
+    }
+  }
+
+  private[mllib]
+  def setLocalHeight(height: Double) = (this.localHeight = height)
+
+  /**
+   * Converts to a adjacency list
+   *
+   * @return List[(fromNodeId, toNodeId, distance)]
+   */
+  def toAdjacencyList(): Array[(Int, Int, Double)] = {
+    val nodes = toArray()
+
+    var adjacencyList = Array.empty[(Int, Int, Double)]
+    nodes.foreach { parent =>
+      if (parent.children.size > 1) {
+        val parentIndex = nodes.indexOf(parent)
+        parent.children.foreach { child =>
+          val childIndex = nodes.indexOf(child)
+          adjacencyList = adjacencyList :+(parentIndex, childIndex, parent.localHeight)
+        }
+      }
+    }
+    adjacencyList
+  }
+
+  /**
+   * Converts to a linkage matrix
+   * Returned data format is fit for scipy's dendrogram function
+   * SEE ALSO: scipy.cluster.hierarchy.dendrogram
```

Comment:
Probably drop this, I think it's unusual to have a `SEE ALSO` if it's from another library.